### ISSUE TITLE
Double printed tab pane headings of the playlist editors

### ISSUE THREAD

**Issue by Borewit** [Borewit/listFix#55]
> I cannot reproduce the issue, opened a large number of playlists, and closed all of those, and then opened then again, repeated that a few times, no side-effects:

We're missing something here. It seems like this issue is due to one or other file corruption. It does work properly with a clean installation, and that *.ini file removed. What causes this issue is yet to be established. It seems to be something that was introduced in later builds.

Just now, I deleted all the json settings files, restarted the app and set it up all over again. Then reopened the 12 playlists for repair, and yet their titles were still contaminated. It looks like two superimposed titles, as I explained before. 

What happened next, is surprising, I opened the Options dialogue, and tried to change the number of closest matches to find. The app immediately started to hang... I moved the indicator around... video:

![image](https://user-images.githubusercontent.com/23255260/217488757-5f9160fa-976c-4d3c-9923-b9d4c26df865.png)

[Animation demonstrating the issue](https://user-images.githubusercontent.com/65730626/216819752-9619d474-35b3-47e9-92a7-1364e03aea89.gif)

_Originally posted by @touwys in https://github.com/Borewit/listFix/issues/51#issuecomment-1417762266_

**Comment by Borewit** [comment]
@touwys Looking to the [animation](https://user-images.githubusercontent.com/65730626/216819752-9619d474-35b3-47e9-92a7-1364e03aea89.gif) it looks like the "GUI-threads" hangs, essentially making you application unresponsive and the rendering (painting) to fail.

**Comment by touwys** [comment]
> @touwys Looking to the [animation](https://user-images.githubusercontent.com/65730626/216819752-9619d474-35b3-47e9-92a7-1364e03aea89.gif) it looks like the "GUI-threads" hangs, essentially making you application unresponsive and the rendering (painting) to fail.

What do I look for, and what can be done about it? Can the graphics card be to blame for it, e.g. driver? Both were recently installed.

**Comment by Borewit** [comment]
> What do I look for, and what can be done about it? Can the graphics card be to blame for it, e.g. driver? Both were recently installed.

Good question. Maybe you cannot do much about it at this moment. By keep improving (refactoring) the code, I hope we  slowly shake out issues. There are still area's many area's where I can improve logging and exception handling. 

I need to focus on some other things now for a few days. But when I have time I will also play around see if I can trigger problems. I can also deploy it do another Windows PC, see if that makes a difference.

I am also adding [unit tests](https://en.wikipedia.org/wiki/Unit_testing), they test smaller software components, like loading / saving playlists. 

The great thing about unit tests are, they can be checked automatically. An each commit I push to a PR, those are run and the PR will turn red if something fails. What makes it tough for this application write unit test, is that the all the file path are OS specific.

**Comment by touwys** [comment]
> Good question. Maybe you cannot do much about it at this moment. By keep improving (refactoring) the code, I hope we slowly shake out issues. There are still area's many area's where I can improve logging and exception handling.

Good, so far the graphics card is working properly, with zero issues manifesting. Installed it 7 days ago.

> I need to focus on some other things now for a few days.

No rush, I need a break too 😂

> I am also adding [unit tests](https://en.wikipedia.org/wiki/Unit_testing), they test smaller software components, like loading / saving playlists.

It looks like these unit tests are going to be a huge productivity boon for weeding out any issues. Are the test units to be run under a single controlling app?

**Comment by Borewit** [comment]
> Are the test units to be run under a single controlling app

Multiple ways to execute the unit tests.

In principle is done via Gradle, running `./gradlew : test`

Which can be executed by my IDE (development environment):
![image](https://user-images.githubusercontent.com/23255260/216835240-b99bca5c-d991-48d6-9ed2-9851876cf8d5.png)

It's done as part of the Pull-Requests (PRs), just added some better visual feedback: https://github.com/Borewit/listFix/actions/runs/4097667916/jobs/7066209974

So basically nothing is able to be merged to the default branch without passing unit testing.

**Comment by Borewit** [comment]
@vall77 now you got [listFix running on Linux](https://github.com/Borewit/listFix/issues/16#issuecomment-1422048343), can you try to reproduce this issue?

**Comment by touwys** [comment]
> @vall77 now got [listFix running on Linux](https://github.com/Borewit/listFix/issues/16#issuecomment-1422048343), can you try to reproduce this issue?

@Borewit: Is this intended for me?

**Comment by Borewit** [comment]
> Is this intended for me?

No, was addressed to @val77. As he has totally different environment, I can shed a different light on this issue.
There is also a risk he finds issue we do not experience, as the path handling is different on Unix.
I have done 0 testing on Unix with the last versions.

**Comment by touwys** [comment]
@Borewit

Any feedback on this test within the Linux environment, yet?

In build **listFix_2.5.1-18** , the issue (of the overlaid tab-text) persists.

**Comment by Borewit** [comment]
Issue still persist in [listFix_2.5.1-19](https://www.mediafire.com/file/ez3yydkdjwbzpqv/listFix_2.5.1-19.exe/file), reported [here](https://github.com/Borewit/listFix/issues/64#issuecomment-1425577156)

> Any feedback on this test within the Linux environment, yet?

Not yet

**Comment by touwys** [comment]
Issue persists in build: [listFix_2.5.1-25](https://github.com/Borewit/listFix/issues/64#issuecomment-1426856406)

**Comment by Borewit** [comment]
I cannot neither reproduce this issue under Linux:

![image](https://user-images.githubusercontent.com/23255260/218465242-b4ea5f69-6cda-4ff6-95c9-7a574ac3ab5f.png)

**Comment by Borewit** [comment]
We know now for sure that the _double printed tab pane headings_ experienced by @touwys is related to:

https://github.com/Borewit/listFix/blob/4feac3b29b065f76f23e3235c098b16ef4e15a2f/src/main/java/listfix/swing/JDocumentTabbedPane.java#L118

as tested in PR https://github.com/Borewit/listFix/pull/94#issuecomment-1428050342.
